driver-core/src/main/com/mongodb/connection/LoggingCommandEventSender.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.connection;
+
+import com.mongodb.MongoCommandException;
+import com.mongodb.diagnostics.logging.Logger;
+import com.mongodb.event.CommandListener;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonReader;
+import org.bson.codecs.RawBsonDocumentCodec;
+import org.bson.json.JsonMode;
+import org.bson.json.JsonWriter;
+import org.bson.json.JsonWriterSettings;
+
+import java.io.StringWriter;
+import java.util.Set;
+
+import static com.mongodb.connection.ProtocolHelper.sendCommandFailedEvent;
+import static com.mongodb.connection.ProtocolHelper.sendCommandStartedEvent;
+import static com.mongodb.connection.ProtocolHelper.sendCommandSucceededEvent;
+import static java.lang.String.format;
+
+class LoggingCommandEventSender implements CommandEventSender {
+    private static final int MAX_COMMAND_DOCUMENT_LENGTH_TO_LOG = 1000;
+
+    private final Set<String> securitySensitiveCommands;
+    private final ConnectionDescription description;
+    private final CommandListener commandListener;
+    private final Logger logger;
+    private final long startTimeNanos;
+    private final CommandMessage message;
+    private final String commandName;
+    private volatile BsonDocument commandDocument;
+
+    LoggingCommandEventSender(final Set<String> securitySensitiveCommands, final ConnectionDescription description,
+                              final CommandListener commandListener, final CommandMessage message,
+                              final ByteBufferBsonOutput bsonOutput, final Logger logger) {
+        this.securitySensitiveCommands = securitySensitiveCommands;
+        this.description = description;
+        this.commandListener = commandListener;
+        this.logger = logger;
+        this.startTimeNanos = System.nanoTime();
+        this.message = message;
+        this.commandDocument = message.getCommandDocument(bsonOutput);
+        this.commandName = commandDocument.getFirstKey();
+    }
+
+    @Override
+    public void sendStartedEvent() {
+        if (loggingRequired()) {
+            logger.debug(
+                    format("Sending command '%s ...' with request id %d to database %s on connection [%s] to server %s",
```

Comment:
I don't like that it prints "..." even when the command wasn't actually truncated.  It could be fixed by having the JsonWriter track whether truncation occurred.

Reply:
Or the getTruncatedJsonCommand method should add the ellipsis, indicating truncatation.

Reply:
Fixed by adding JsonWriter.isTruncated() method